flowmachine/flowmachine/core/server/action_handlers.py

```diff
@@ -0,0 +1,327 @@
+# handler must be of type `ZMQReply`.
+#
+# In order to actually perform an action, call the `perform_action()`
+# function with the action name and parameters. This selects the correct
+# action handler and also gracefully handles any potential errors.
+#
+
+from apispec import APISpec
+from apispec_oneofschema import MarshmallowPlugin
+from marshmallow import ValidationError
+
+from flowmachine.core import Query, GeoTable
+from flowmachine.core.cache import get_query_object_by_id
+from flowmachine.core.query_info_lookup import (
+    QueryInfoLookup,
+    UnkownQueryIdError,
+    QueryInfoLookupError,
+)
+from flowmachine.core.query_state import QueryStateMachine, QueryState
+from flowmachine.utils import convert_dict_keys_to_strings, sort_recursively
+from .exceptions import FlowmachineServerError
+from .query_schemas import FlowmachineQuerySchema
+from .zmq_helpers import ZMQReply
+
+__all__ = ["perform_action"]
+
+
+def action_handler__ping():
+    """
+    Handler for 'ping' action.
+
+    Returns the message 'pong'.
+    """
+    return ZMQReply(status="success", msg="pong")
+
+
+def action_handler__get_available_queries():
+    """
+    Handler for 'get_available_queries' action.
+
+    Returns a list of available flowmachine queries.
+    """
+    available_queries = list(FlowmachineQuerySchema.type_schemas.keys())
+    return ZMQReply(status="success", payload={"available_queries": available_queries})
+
+
+def action_handler__get_query_schemas():
+    """
+    Handler for the 'get_query_schemas' action.
+
+    Returns a dict with all supported flowmachine queries as keys
+    and the associated schema for the query parameters as values.
+    """
+    spec = APISpec(
+        title="FlowAPI",
+        version="1.0.0",
+        openapi_version="3.0.2",
+        plugins=[MarshmallowPlugin()],
+    )
+    spec.components.schema("FlowmachineQuerySchema", schema=FlowmachineQuerySchema)
+    schemas_spec = spec.to_dict()["components"]["schemas"]
+    schemas_spec = sort_recursively(schemas_spec)  # needed for reproducible output
+    return ZMQReply(status="success", payload={"query_schemas": schemas_spec})
+
+
+def action_handler__run_query(**action_params):
+    """
+    Handler for the 'run_query' action.
+
+    Constructs a flowmachine query object, sets it running and returns the query_id.
+    For this action handler the `action_params` are exactly the query kind plus the
+    parameters needed to construct the query.
+    """
+    try:
+        query_obj = FlowmachineQuerySchema().load(action_params)
+    except ValidationError as exc:
+        # The dictionary of marshmallow errors can contain integers as keys,
+        # which will raise an error when converting to JSON (where the keys
+        # must be strings). Therefore we transform the keys to strings here.
+        error_msg = "Parameter validation failed."
+        validation_error_messages = convert_dict_keys_to_strings(exc.messages)
```

Comment:
Boggles my mind a little that this is necessary. Silly marshmallow :P

Reply:
Yeah... Maybe I missed something how to work around this, but by default marshmallow sometimes exports error messages with an integer as the dict key (presumably because it refers to the index in a list). If we can would be good to find a way to tidy this up a little - in particular the forwarding of error messages received from marshmallow. While they are accurate and reflect all the validation errors, they are a bit inconvenient to interpret from the user's point of view, so we probably want to massage them a little before sending them back.